pkg/rbac/parser.go

```diff
@@ -14,46 +14,34 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package rbac contain libraries for generating RBAC manifests from RBAC
-// annotations in Go source files.
 package rbac
 
 import (
 	"log"
 	"strings"
 
 	rbacv1 "k8s.io/api/rbac/v1"
-	"sigs.k8s.io/controller-tools/pkg/internal/general"
+	"sigs.k8s.io/controller-tools/pkg/internal/annotation"
 )
 
 type parserOptions struct {
 	rules []rbacv1.PolicyRule
 }
 
-// parseAnnotation parses RBAC annotations
-func (o *parserOptions) parseAnnotation(commentText string) error {
-	for _, comment := range strings.Split(commentText, "\n") {
-		comment := strings.TrimSpace(comment)
-		if strings.HasPrefix(comment, "+rbac") {
-			if ann := general.GetAnnotation(comment, "rbac"); ann != "" {
-				o.rules = append(o.rules, parseRBACTag(ann))
-			}
-		}
-		if strings.HasPrefix(comment, "+kubebuilder:rbac") {
-			if ann := general.GetAnnotation(comment, "kubebuilder:rbac"); ann != "" {
-				o.rules = append(o.rules, parseRBACTag(ann))
-			}
-		}
-	}
-	return nil
+func (o *parserOptions) AddToAnnotation(a annotation.Annotation) annotation.Annotation {
+	a.Module(&annotation.Module{
+		Name: "rbac",
+		Do:   o.parseRBAC,
+	})
+	return a
```

Comment:
Good point ! This PR hasn't well illustrated use of annotation.

The ideal `Modularized Annotation Pattern` should be like : 
1) Initialize a default, global annotation
2) Register module with its handler function and context into global annotation instance
3) Call general Parse functions to invoke parsing dir, files, go-types and their annotations . If encounter a validate module registered in annotation, invoking `Do` handler function of module or submodule. 

The sample code in this PR is not idea, because I just reuse the old functions for demo. `Do` function in Module could invoke `types.Type`, `APIResource`, and other runtime context. There should be more design for complex scenarios.

Module struct has a `Meta` field for holding context that will be impacted
```golang
// Module defines functional feature for annotation. Header may contain multiple modules,
// single module may contain submodules. Coresponding Module invokes Do function when valid module name is found in parsing annotations
type Module struct {

	// Name of the module. It should match the token string in the annotation
	Name string
	// Meta holds meta data this module will return or impact. (TODO: fanz) Consider take it into context
	Meta interface{}
	// SubModules represents a recursive architecture of annotation syntax, e.g. [header]:[module]:[submodule1]:[submodule2]:...
	SubModules map[string]*Module
	// Do is handler function which defines what this module can do. It takes annotation token passed by Module, and might involve context from runtime
	Do func(string) error
}

```